CSharp/Library/Microsoft.Bot.Connector.Shared/MicrosoftAppCredentials.cs

```diff
@@ -266,6 +291,13 @@ private bool TokenNotExpired(OAuthResponse token)
             return token.expiration_time > DateTime.UtcNow;
         }
 
+        private bool TokenHalfwayExpired(OAuthResponse token, int secondsToHalfwayExpire = 1800, int secondsToExpire = 60)

```

Comment:
>1800 [](start = 91, length = 4)

this should be dynamically calculated from OAuthResponse.expires_in value. 

probably: Math.Min(OauthResponse.expires_in / 2, 1800)